torchao/quantization/quantize_/float8/float8_tensor.py

```diff
@@ -0,0 +1,600 @@
+        activation_granularity: Optional[FP8Granularity] = None,
+        activation_scale_ub: Optional[float] = None,
+        mm_config: Optional[Float8MMConfig] = None,
+        gemm_kernel_choice: str = GemmKernelChoice.ATEN,
+    ):
+        assert activation_dtype in [None, torch.float8_e4m3fn, torch.float8_e4m3fnuz]
+        block_size = list(block_size)
+
+        if activation_scale_ub is None:
+            activation_scale_ub = 1200.0
+
+        activation_scale_ub = torch.tensor(
+            [activation_scale_ub],
+            dtype=torch.float,
+            device=w.device,
+        )
+
+        if gemm_kernel_choice == GemmKernelChoice.FBGEMM:
+            assert isinstance(activation_granularity, PerRow), (
+                f"Only per row activation is supported for {gemm_kernel_choice} got {activation_granularity}"
+            )
+            assert weight_dtype == activation_dtype
+            wq, w_scale = torch.ops.fbgemm.quantize_fp8_per_row(w)
+            # add a last dimension for per row quantization to align the rank of
+            # w_scale and wq
+            w_scale = w_scale.unsqueeze(-1).contiguous()
+            dtype = w.dtype
+            del w
+            return Float8Tensor(
+                wq,
+                w_scale,
+                activation_scale_ub=activation_scale_ub,
+                block_size=block_size,
+                activation_dtype=activation_dtype,
+                activation_granularity=activation_granularity,
+                mm_config=mm_config,
+                gemm_kernel_choice=gemm_kernel_choice,
+                dtype=dtype,
+            )
+
+        else:
+            w_scale = _choose_qparams_affine_float8(
+                w, float8_dtype=weight_dtype, block_size=block_size
+            )
+            wq = _quantize_affine_float8(w, w_scale, weight_dtype)
+            dtype = w.dtype
+            del w
+            return Float8Tensor(
+                wq,
+                w_scale,
+                activation_scale_ub=activation_scale_ub,
+                block_size=block_size,
+                activation_dtype=activation_dtype,
+                activation_granularity=activation_granularity,
+                mm_config=mm_config,
+                gemm_kernel_choice=gemm_kernel_choice,
+                dtype=dtype,
+            )
+
+
+implements = Float8Tensor.implements
+
+
+@implements([torch.nn.functional.linear, aten.linear.default])
+def _(func, types, args, kwargs):
+    input_tensor, weight_tensor, bias = (
+        args[0],
+        args[1],
+        args[2] if len(args) > 2 else None,
+    )
+    orig_act_size = input_tensor.size()
+    orig_out_features = weight_tensor.shape[-2]
+
+    if weight_tensor.gemm_kernel_choice == GemmKernelChoice.FBGEMM:
+        # not used
+        num_tokens = torch.empty([input_tensor.size(0)], device=input_tensor.device)
+        a_data, a_scale = torch.ops.fbgemm.quantize_fp8_per_row(
```

Comment:
if we want to use `fbgemm`'s quantize, perhaps the enum should be named for "backend"?

do we even want to hardcode the use of the dedicated quantize kernel here? I feel like the optimal solution should be:
0. ensure fbgemm quantize kernel and native pytorch quantize kernel are bitwise equivalent
1. use fbgemm quantize kernel in eager mode
2. use whatever is faster between fbgemm kernel and torch.compile if compile is on

for short term hardcoding the kernel here seems ok, but can we design the system in a way to enable us to do ^ in the future without breaking BC?

Reply:
sounds good to align the numerics, although not sure if it's possible, I have created a issue to track: https://github.com/pytorch/ao/issues/2511

I can rename gemm_kernel_choice to Backend, if people are OK with have this as an explicitly top level concept